File: Spring 2021/airtable/starter-code/ruby/sql_evaluator.rb
# frozen_string_literal: true

require "json"

def main
  if ARGV.length != 3
    $stderr.puts "Usage: COMMAND <table-folder> <sql-json-file> <output-file>"
    exit 1
  end

  table_folder, sql_json_file, output_file = ARGV

  # Load the query JSON.
  begin
    query = load_json_from_file(sql_json_file)
  rescue JsonFileLoadError => e
    $stderr.puts "Error loading SQL JSON from #{q(sql_json_file)}: #{e}"
    exit 1
  end

  # Load all the tables referenced in the "FROM" clause.
  tables = []
  query["from"].each do |table_decl|
    table_source_path = File.join(table_folder, "#{table_decl["source"]}.table.json")
    raw_table = load_json_from_file(table_source_path)
    begin
      raw_table = load_json_from_file(table_source_path)
    rescue JsonFileLoadError => e
      $stderr.puts "Error loading table JSON from #{q(sql_json_file)}: #{e}"
      exit 1
    end
    table = Table.new(raw_table[0], raw_table.drop(1))
    tables.push(table)
  end

  # TODO: Actually evaluate query.
  # For now, just dump the input back out.
  File.open(output_file, "w") do |out|
    write_json_indented(out, query)
    tables.each do |table|
      write_table(out, table)
    end
  end
end

class Table
  def initialize(columns, rows)
    @columns = columns
    @rows = rows
  end

  attr_reader :columns, :rows
end

class JsonFileLoadError < StandardError
end

def load_json_from_file(path)
  contents = File.read(path, encoding: "ascii")
  JSON.parse(contents)
rescue SystemCallError, IOError => e
  raise JsonFileLoadError, "Error reading: #{e}"
rescue JSON::ParserError => e
  raise JsonFileLoadError, "Invalid JSON: #{e}"
end

def write_table(out, table)
  out << "[\n"
  out << "    #{JSON.generate(table.columns)}"
  table.rows.each do |row|
    out << ",\n    #{JSON.generate(row)}"
  end
  out << "\n]\n"
end

def write_json_indented(out, value)
  json = JSON.pretty_generate(value, indent: "    ")
  # For readability, remove the newlines and whitespace that Ruby's JSON formatter
  # adds to empty arrays and objects.
  json = json.gsub(/{\n *}/m, "{}").gsub(/\[\n *\n *\]/m, "[]")
  out << json
  out << "\n"
end

# For quoting strings in messages to the user.
def q(s)
  s.to_json
end

if __FILE__ == $PROGRAM_NAME
  main
end

File: Spring 2021/airtable/starter-code/python-typed/sql_evaluator.py
from __future__ import annotations

from typing import Any, List, TextIO

from model import ColumnDef, SqlValue, Query

import json
import os
import sys

assert sys.version_info >= (3, 8), (
    "This program requires Python 3.8+.  "
    "You're trying to run it with Python {}.".format(".".join(map(str, sys.version_info[:3]))))

def main() -> None:
    args = sys.argv[1:]
    if len(args) != 3:
        sys.stderr.write("Usage: COMMAND <table-folder> <sql-json-file> <output-file>\n")
        sys.exit(1)
    table_folder, sql_json_file, output_file = args

    # Load the query JSON.
    try:
        query: Query = load_json_from_file(sql_json_file)
    except JsonFileLoadError as e:
        sys.stderr.write("Error loading SQL JSON from {}: {}\n".format(q(sql_json_file), e))
        sys.exit(1)

    # Load all tables referenced in the "FROM" clause.
    tables = []
    for table_decl in query["from"]:
        table_source_path = os.path.join(table_folder, "{}.table.json".format(table_decl["source"]))
        try:
            raw_table = load_json_from_file(table_source_path)
        except JsonFileLoadError as e:
            sys.stderr.write("Error loading table JSON from {}: {}\n".format(q(table_source_path), e))
            sys.exit(1)
        table = Table(raw_table[0], raw_table[1:])
        tables.append(table)

    # TODO: Actually evaluate query.
    # For now, just dump the input back out.
    with open(output_file, "w") as out:
        json.dump(query, out, indent=4)
        out.write("\n")

        for table in tables:
            write_table(out, table)

class Table:
    columns: List[ColumnDef]
    rows: List[List[SqlValue]]

    def __init__(self, columns: List[ColumnDef], rows: List[List[SqlValue]]):
        self.columns = columns
        self.rows = rows

class JsonFileLoadError(Exception):
    pass

def load_json_from_file(path: str) -> Any:
    try:
        with open(path, "r", encoding="ascii") as f:
            return json.load(f)
    except IOError as e:
        raise JsonFileLoadError("Error reading: {}".format(e))
    except ValueError as e:
        raise JsonFileLoadError("Invalid JSON: {}".format(e))

def write_table(out: TextIO, table: Table) -> None:
    out.write("[\n")
    out.write("    {}".format(json.dumps(table.columns, separators=(",", ":"))))
    for row in table.rows:
        out.write(",\n    {}".format(json.dumps(row, separators=(",", ":"))))
    out.write("\n]\n")

# For quoting strings in messages to the user
def q(s: str) -> str:
    return json.dumps(s)

if __name__ == "__main__":
    main()
